### ISSUE TITLE
Not Saving Custom Ships Correctly.

### ISSUE THREAD

**Issue by curlymoe** [goatfungus/NMSSaveEditor#981]
When I export a custom built ship, and then restore it, I don't get the same ship back, just a random ship of the same group ie hauler.
I have tried using the json editor to copy the custom ship and paste it into another slot, and I still get another random ship. I have tried comparing json from a custom ship, and a standard one, and they look the same to my untrained eye. But obviously there is something different about the custom ships in the save that the Editor is not picking up.

 Many thanks for all your work on this Editor, and for updating so quickly.

**Comment by Ymmit34** [comment]
Having the exact same problem; whenever I export and import a custom ship it gives me a random one and resets to being an "authenticated" ship.

**Comment by goatfungus** [comment]
I think there's something new in the ship data that isn't present in the backups. Not sure what it is yet, but working on it.

**Comment by Nobere** [comment]
I changed the seed of a created ship and it was the same when I reloaded into the game.

I think the custom-built ships no longer use a seed to determine how they look. 
I’m betting there is something on the save file that saves their appearance.

**Comment by EdVenture2025** [comment]
When I import any sentinel ship it looks like the same plain orange ship.  If I edit a constructed ship it changes to authenticated, but looks different.  I'm guessing it either the seed thing Nobere mentioned or the authentication bit needs to be left off.

**Comment by curlymoe** [comment]
> When I import any sentinel ship it looks like the same plain orange ship. If I edit a constructed ship it changes to authenticated, but looks different. I'm guessing it either the seed thing Nobere mentioned or the authentication bit needs to be left off.

Yes, I think this may be the key. We need to find out how a ship gets authenticated or not.

**Comment by boldfish** [comment]
Can I add here that once you have exported and reimported a custom built ship it can then make a muddle of your other ships, they change to something they weren't.  I have had shuttles change shape and color, exotics become a completely different type and solar ships change color and type too. I've imported the same saved non custom ship twice on one save and with the same seed get two different models of shuttle, so whatever is happening when you export and re-import a custom ship it affects more than the custom ship and messes up the entire save.

I've scrapped all imported custom ships, reimported normal saves and seem to have my save back.

So beware of export and import on custom ships!!!

**Comment by tana-second** [comment]
Custom ships appear to save additional information in `BaseContext/PlayerStateData/CharacterCustomizationData` for a non-Expedition save. This is a 23-object array where objects 3-8 govern ship slots 1-6 and objects 17-22 govern ship slots 7-12. I could copy the Iron Vulture from one save file to another via copying the ship data normally, then using the raw JSON editor and manually copying the contents of object 7 (the Vulture was the fifth ship acquired) from the first save and replacing the contents of object 17 (the imported ship was the seventh ship) in the destination save. I was able to repeat the same type of operation using a customized Explorer, so it's not specific to the Vulture.

For reference, this was in the array object corresponding to the Vulture:

```json
{
	"SelectedPreset":"^",
	"CustomData":{
		"DescriptorGroups":[
			"^DROPS_COCKS13",
			"^DROPS_ENGIS13",
			"^DROPS_WINGS13"
		],
		"FCx":"^SHIP_METALLIC",
		"Colours":[
			{
				"Palette":{
					"Palette":"Paint",
					"ColourAlt":"Primary"
				},
				"Colour":[
					0.4572489857673645,
					0.4572489857673645,
					0.4572489857673645,
					1.0
				]
			},
			{
				"Palette":{
					"Palette":"Paint",
					"ColourAlt":"Alternative3"
				},
				"Colour":[
					1.0,
					1.0,
					1.0,
					1.0
				]
			},
			{
				"Palette":{
					"Palette":"Undercoat",
					"ColourAlt":"Primary"
				},
				"Colour":[
					0.8558139801025391,
					0.5949479937553406,
					0.4502660036087036,
					1.0
				]
			}
		],
		"TextureOptions":[
			{
				"TextureOptionGroupName":"^SHIP_FIGHT",
				"TextureOptionName":"^METALBOLT"
			}
		],
		"BoneScales":[],
		"Scale":1.0
	}
}
```
I suspect that if you try to copy a ship into a slot with customization information, weird or bad things can happen if there's a ship type mismatch. You will need to reset the customization data in that case with default data:

```json
{
	"SelectedPreset":"^",
	"CustomData":{
		"DescriptorGroups":[],
		"FCx":"^",
		"Colours":[],
		"TextureOptions":[],
		"BoneScales":[],
		"Scale":1.0
	}
}
```

**Comment by boldfish** [comment]
Thank you.  

The issue I believe, is that when deleting a custom ship using the save editor, the customisation on the slot is not removed.

I can confirm that setting slots back to defaults after deleting custom ships does fix the problem.

Deleting custom ships and importing new ships, or adding new ships in game to those previously customised slots breaks everything.

So

1. don't delete a custom ship in the save editor, unless you also edit the JSON to remove the customised bits above and set it back to default.
2. don't rely on the ship order in the save editor drop down to know which slots to edit in the JSON - you need to check ship ownership section in the JSON to get the correct slot numbers. (and ships are numbered 0 - 11)
3. Due to slot numbers starting at 0, Your first ship is in slot 0 and is controled by object 3;  slot 1 is object 4 etc...  the ship in slot 6 is controlled by object 17

**Comment by boldfish** [comment]
> I think there's something new in the ship data that isn't present in the backups. Not sure what it is yet, but working on it.

As tana-second points out there is extra data for custom ships in BaseContext/PlayerStateData/CharacterCustomizationData

Current exports don't include it, so you can't import a custom ship, you just get the unskinned base model

Deleting a custom ship using the save editor, leaves it behind and will corrupt the ship that later occupies that slot, either in game or imported.

Interestingly it might open up opportunities to reskin existing ships?

@goatfungus I've joined your patreon subs in the hope that you might find a way...

:O)

**Comment by SalemDarling** [comment]
I was having this problem to, I figured it must be in some other data but I forgot about the Character CustomizationData. It used to be that changes to that section only showed for yourself but not to other players. Maybe they changed it now? In which case you can put in any RGB values you want. At least I hope they changed it, it would bad if we are the only ones who can see the correct colours on our custom ships lol

**Comment by SalemDarling** [comment]
> Custom ships appear to save additional information in `BaseContext/PlayerStateData/CharacterCustomizationData` for a non-Expedition save. This is a 23-object array where objects 3-8 govern ship slots 1-6 and objects 17-22 govern ship slots 7-12. I could copy the Iron Vulture from one save file to another via copying the ship data normally, then using the raw JSON editor and manually copying the contents of object 7 (the Vulture was the fifth ship acquired) from the first save and replacing the contents of object 17 (the imported ship was the seventh ship) in the destination save. I was able to repeat the same type of operation using a customized Explorer, so it's not specific to the Vulture.
> 
> For reference, this was in the array object corresponding to the Vulture:
> 
> ```json
> {
> 	"SelectedPreset":"^",
> 	"CustomData":{
> 		"DescriptorGroups":[
> 			"^DROPS_COCKS13",
> 			"^DROPS_ENGIS13",
> 			"^DROPS_WINGS13"
> 		],
> 		"FCx":"^SHIP_METALLIC",
> 		"Colours":[
> 			{
> 				"Palette":{
> 					"Palette":"Paint",
> 					"ColourAlt":"Primary"
> 				},
> 				"Colour":[
> 					0.4572489857673645,
> 					0.4572489857673645,
> 					0.4572489857673645,
> 					1.0
> 				]
> 			},
> 			{
> 				"Palette":{
> 					"Palette":"Paint",
> 					"ColourAlt":"Alternative3"
> 				},
> 				"Colour":[
> 					1.0,
> 					1.0,
> 					1.0,
> 					1.0
> 				]
> 			},
> 			{
> 				"Palette":{
> 					"Palette":"Undercoat",
> 					"ColourAlt":"Primary"
> 				},
> 				"Colour":[
> 					0.8558139801025391,
> 					0.5949479937553406,
> 					0.4502660036087036,
> 					1.0
> 				]
> 			}
> 		],
> 		"TextureOptions":[
> 			{
> 				"TextureOptionGroupName":"^SHIP_FIGHT",
> 				"TextureOptionName":"^METALBOLT"
> 			}
> 		],
> 		"BoneScales":[],
> 		"Scale":1.0
> 	}
> }
> ```
> 
> I suspect that if you try to copy a ship into a slot with customization information, weird or bad things can happen if there's a ship type mismatch. You will need to reset the customization data in that case with default data:
> 
> ```json
> {
> 	"SelectedPreset":"^",
> 	"CustomData":{
> 		"DescriptorGroups":[],
> 		"FCx":"^",
> 		"Colours":[],
> 		"TextureOptions":[],
> 		"BoneScales":[],
> 		"Scale":1.0
> 	}
> }
> ```

Interesting that those colours are 117 grey, white, and orange, and I accidentally got that data onto a custom fighter of mine. It didn't crash anything, even with the hauler parts listed, it just totally changed the cockpit, wings, and engine configuration to another fighter type but still with the grey, white, and orange custom colours lol

**Comment by brooksie68** [comment]
Not a dev, I'm a ux guy. But after designing enterprise software for the last 20 years I know what JSON is and I can look at it and make sense of it, see what it's doing.

And my ship saves are alllll kind of jacked up. I deleted a ship w the Save Editor and I scrapped a few at the Space Station and somewhere in there things started getting really weird.

My starter ship became a hulking Sentinel ship. The naming of my ships seems off... Whatever is in slot 0 takes the name of my starter ship. And the last w slots keep filling with random ships I didn't bug or add.

What I am wondering is if I can get the raw JSON for the starter ship with nothing and use that to replace what's in slot 0 and get my starter ship back?

And also, maybe I should just salvage all my ships and start over?

IDK, I'm sad, I have almost 400 hours in the game and I don't really want to start over. I've built a couple of huge bases that I really love.

But my ship situation is majorly screwed up and I'm not sure the best way to fix it.

No complaint about the editor though... Thing is AWESOME and I appreciate the work you're doing here very much. Also, the devs at my job got a huuuge kick out of your username.

**Comment by brooksie68** [comment]
Just cheking in again to say that I solved my problem by just restoring from an earlier save before I starting mucking about with deleting ships in the editor. So I got myself into the problem with the editor and then got myself back out with it as well. VERY thoughtful of Mr goatfungus to put that auto archiving feature in there. I'm gonna have to go over to Patreon and make a donation. Thanks!

**Comment by RetiredNASAnerd** [comment]
It would seem that *always* exporting/importing all of the associated ship data from BaseContext/PlayerStateData/CharacterCustomizationData, along with what is already being saved, would be the simplest fix for this issue. Not only would it correctly archive/restore custom ships, it would also do so for the expedition reward ships. Lastly, this would also fix issues with corrupted ship slots where custom ships were deleted in the save editor.  Obviously you can't recreate missing data for older ship save files (.SH0), so I'd also suggest setting the filetype of the new save files to (.SH1). Simply saving everything in that structure would, to a large extent, futureproof the save editor from HG adding things there.

**Comment by RetiredNASAnerd** [comment]
So I inadvertently did a little test and can confirm the solution to this issue.  I stupidly deleted the second ship in the ship list with the save editor. Apparently, the save editor moves the ships below the deleted ship up one in the array. Several of the ships below that deleted ship changed. My Iron Vulture (#3) became an ugly hauler. My Boundary Herald (#4) changed colors. And an Explorer changed some structure details and color. The rest of the ships further down the list were all regular ships - neither custom or reward ships - so no special data was in those elements of CharacterCustomizationData array. To try and fix the problem, I duped the save file and looked at the raw JSON. The save editor moved the ships up the list, but did not alter the list in CharacterCustomizationData. So I Simply copy/pasted everything in [5] to [4], [6] to [5], and duped [7] to [6].  Restarting the game with that save showed that all of the ships were back to normal. Whew ! There is zero doubt that archiving/restoring the data from CharacterCustomizationData along with the normally saved ship data will solve this problem.

**Comment by wkitty42** [comment]
excellent detective work!

now the question is if the deleted slot should simply be empty thus maintaining the index ordering and spacing or if CharacterCustomizationData should also be modified...

what does NMS do if you delete the ship in the 2nd slot from within the game? does the 2nd slot and its CCD simply go blank or are the other ships and their CCD moved up to fill in the hole?

**Comment by RetiredNASAnerd** [comment]
Yes, the game simply empties the slot.  The game also fills them from the top down, so that newly emptied slot will be the first to be refilled.

**Comment by wkitty42** [comment]
that's what i thought... so with the new stuff, when deleting a ship, not only is the ship slot left empty but also the matching CCD slot...

**Comment by xXGlitchI0Xx** [comment]
Ok, so from my understanding you can't just import ships then. Can anyone give me a step-by-step on how to import the iron vulture to a save that I'm modding using goatfungus.

**Comment by wkitty42** [comment]
> Ok, so from my understanding you can't just import ships then. Can anyone give me a step-by-step on how to import the iron vulture to a save that I'm modding using goatfungus.

you have to have a craft that was exported to be able to import it... to add it this way, someone with the craft would have to use this editor to export the craft to a file, send you that file, and you then import it into your save...

it is much easier to enable it in Account->Season Rewards... then go to the Anomaly and claim it from the QS guy...

**Comment by RetiredNASAnerd** [comment]
A long post by tana-second earlier in this thread had the missing bits for the Iron Vulture that need to be added to the corresponding slot in the CCD array. If that does not make sense, then you probably don't want to be editing the raw JSON. If you have already earned the Iron Vulture in another save, just go and claim it again at the Nexus in the other save(s).

**Comment by RetiredNASAnerd** [comment]
> 

Most ships export and import just fine. Ships you have fabricated and some expedition reward ships do not export/import correctly. FYI: the release of the Starborn Phoenix introduced a new "version" of the Starborn Runner and both the Phoenix and new Runner can be exported and imported without issue.

**Comment by SalemDarling** [comment]
This really isn't rocket science ... okay well maybe it is lol :)  The attached text file is the Iron Vulture ship in all it's tricked out glory. The attached images will show you the stats you get if you use this file. Copy the text file and paste it in the ship slot of your choice under ShipOwnership as shown in the images.
 
 Step 2 is you have to use the excellent data provided above by tana-second for the CCD and put it in the slot corresponding to the one you used for your Iron Vulture, under CharacterCustomizationData, which is located a little below ShipOwnership as shown in the images.  

I maxxed this one out to show what can be done. Not much is going to phase you with these stats, the tech slots are all energized, and you have over 22,000 LY warp distance.

[IRON VULTURE.txt](https://github.com/user-attachments/files/18284528/IRON.VULTURE.txt)
![IV01](https://github.com/user-attachments/assets/d7debeab-fc43-40c3-9466-de24b1956305)
![IV02](https://github.com/user-attachments/assets/0e10db0b-5f89-40f5-a17b-b4d14c88df8b)
![IV03](https://github.com/user-attachments/assets/811dec8e-9dab-4886-8e86-2512d9a4f5ee)
![IV04](https://github.com/user-attachments/assets/98dba2d6-f545-4625-95aa-51b407cb63b2)
![IV05](https://github.com/user-attachments/assets/5abcb7e4-7796-4238-9b2c-859e8f9b23c5)

**Comment by RetiredNASAnerd** [comment]
You are missing the bigger picture. This isn't just about the Iron Vulture. It is about most expedition reward ships and all custom built ships. It is about being unable to delete a ship correctly if any of the above mentioned ships are owned by the player further down their ship list. It is about being unable to export/restore said ships. We can't (easily) move custom/reward ships between saves. Fortunately HG gave us 18 new places to store these ships while we await a fix for this (very fixable) issue.

**Comment by Geldric** [comment]
Not sure “who” you think is missing some bigger picture… since mine is the last comment, I’ll presume it was me. Apologies if it was directed towards someone else.

The Iron Vulture is the only expedition reward ship that requires this extra step. All other reward ships can be exported & imported using the editor just fine.

You also mentioned “built” ships & this process 100% applies to them. Pretty sure I’ve grasped the correct picture size 👍

**Comment by GlitchGMG** [comment]
> Not sure “who” you think is missing some bigger picture… since mine is the last comment, I’ll presume it was me. Apologies if it was directed towards someone else.
> 
> The Iron Vulture is the only expedition reward ship that requires this extra step. All other reward ships can be exported & imported using the editor just fine.
> 
> You also mentioned “built” ships & this process 100% applies to them. Pretty sure I’ve grasped the correct picture size 👍

To be honest I don't know why this is in here anymore I just created a save on PC NMS and transferred all the JSON to a PS4 save and it transferred just fine. Haven't had an issue with transferring the new wraith either.

**Comment by wkitty42** [comment]
> To be honest I don't know why this is in here anymore I just created a save on PC NMS and transferred all the JSON to a PS4 save and it transferred just fine.

FWIW: if you go all the way back to the original post, it was about exporting and importing ships... like if you are moving them around in your list to sort or put them in a specific order OR if you are copying them between saves _on the same device_... cross-saves weren't yet a thing and yes, everything should transfer correctly between them which is why is is currently undergoing a long beta testing cycle with live data...

the reason this issue isn't closed, yet, is because it was during the time "we" were digging about in it, the additional CharacterCustomizationData block was found... it is a separate key from the ship and contains the additional special data for it... now that it is known, hopefully goatfungus can use the info to fix the export/import of the ships that have this additional CharacterCustomizationData block...

**Comment by Geldric** [comment]
⬆️ @wkitty42 is correct, the issue still stands for the Iron Vulture and all Traveller-built ships.

Now every time I start (yet another new) Titan save, I can just drop the IV data into CharacterCustomizationData and by the time I get to my ship, she's just waiting there smiling at me... yay!

![Image](https://github.com/user-attachments/assets/7f9aab0e-02ee-4159-a7fd-5bed5d29bf13)
